Zend/tests/function_arguments/sensitive_parameter_value_unserialize.phpt

```diff
@@ -0,0 +1,16 @@
+--TEST--
+The SensitiveParameterValue may not be unserialized.
+--FILE--
+<?php
+
+$v = new SensitiveParameterValue('secret');
+
+var_dump(unserialize(serialize($v)));
+
+--EXPECTF--
+Fatal error: Uncaught Exception: Unserializing objects of class SensitiveParameterValue is not supported. in %ssensitive_parameter_value_unserialize.php:5
```

Comment:
When writing the RFC my reasoning was that the `SensitiveParameterValue` should “just work” with existing code without needing special casing: Thus allowing serialization, but omitting the inner value.

Because this serialization is a lossy operation, running the deserialization is an undefined operation, thus it is prevented.

But now that you mention it: An exception during unserialization is not recoverable.

I see two options:

1. Disallow both serialization and unserialization.
2. Allow unserialization, but poison the unserialized object and disallow calling `->getValue()` on it.

Reply:
I think I'd prefer option 2 since serializing exceptions can be useful, I think Symfonys profiler makes use of that, can you confirm @nicolas-grekas? But maybe this should be mentioned on the mailing list.

Reply:
Honestly I missed this part of the RFC. Being able to serialize but not unserialize looks extremely WTF to me.
I'm definitely prefer throwing on both sites.

Reply:
> But maybe this should be mentioned on the mailing list.

Mailing list thread: https://externals.io/message/117136

Reply:
> https://www.php.net/manual/en/migration80.incompatible.php

[debug_backtrace()](https://www.php.net/manual/en/function.debug-backtrace.php) and [Exception::getTrace()](https://www.php.net/manual/en/exception.gettrace.php) will no longer provide references to arguments. It will not be possible to change function arguments through the backtrace.

That's important to keep in mind when writing code to handle SensitiveParameterValue or polyfill SensitiveParameterValue in a framework.

Anything looking for SensitiveParameterValue wouldn't find anything prior to php 8.2, though, so that isn't a concern.

I'd prefer 2 since there may be existing/new code that serializes somewhere, e.g.

- checking if exceptions are loosely equivalent
- estimating memory usage of an exception being logged
- unit testing, rare code paths for error handling that unexpectedly call serialize on something containing a trace, etc

I'd consider either better than throwing only on unserialize, I'd missed that earlier.

Reply:
As per the discussion on the ML (unanimous "disallow completely") and the comments here in the PR, I've adjusted the implementation to completely disallow serialization. As per @Danack:

> "no is temporary, yes is forever"

Thanks for voicing your opinion on this matter. I'll make sure to update the RFC with an errata section later.